### DISCUSSION TITLE
如何设置MPV使用系统代理？

### DISCUSSION THREAD

**Discussion by HorrZzz** [hooke007/mpv_PlayKit#93] [mpv 通用]
请问如何设置MPV使用系统代理？
还有能否显示实时网速？

**Comment by hooke007** [comment]
它。。只是个本地播放器啊

使用系统代理和显示实时网速不应该是其他软件的职能吗...

**Reply by hooke007** [reply]
不清楚你是不是指ytdl的代理 https://github.com/ytdl-org/youtube-dl#network-options
ytdl的参数mpv通过这个参数传递 https://github.com/hooke007/MPV_lazy/blob/1c10309bc9cc80a36707d92b5062456be0fef890/portable_config/mpv.conf#L192

印象中并没有提供直接对应缓存速度的属性

**Comment by snylonue** [comment]
代理可以设置 [HTTP_PROXY](https://mpv.io/manual/master/#environment-variables-http-proxy) 或者 [--http-proxy](https://mpv.io/manual/master/#options-http-proxy) 但是测试下来不太正常
实时网速的话 stats 第三页里有写，也有对应的 property: [cache-speed](https://mpv.io/manual/master/#command-interface-cache-speed)

**Comment by ptr7F** [comment]
ffmpeg好像不支持sock5，目前用http代理正常。
Windows下设置环境变量。假设代理地址为127.0.0.1，端口为1234。
cmd:
set HTTP_PROXY=http://127.0.0.1:1234
pwsh:
$env:HTTP_PROXY="http://127.0.0.1:1234"
以上只是临时变量，要长期有效且不用命令行可以设置永久环境变量（是系统变量还是用户变量都可以）
方法一：
在按下windows开始按钮输入“env”，进入系统属性，然后点击“环境变量”，新建一个变量，变量名为HTTP_PROXY，变量值为http://127.0.0.1:1234
方法二：
管理员权限运行CMD：
setx HTTP_PROXY http://127.0.0.1:1234

**Reply by HorrZzz** [reply]
> ffmpeg好像不支持sock5，目前用http代理正常。 Windows下设置环境变量。假设代理地址为127.0.0.1，端口为1234。 cmd: set HTTP_PROXY=http://127.0.0.1:1234 pwsh: $env:HTTP_PROXY="http://127.0.0.1:1234" 以上只是临时变量，要长期有效且不用命令行可以设置永久环境变量（是系统变量还是用户变量都可以） 方法一： 在按下windows开始按钮输入“env”，进入系统属性，然后点击“环境变量”，新建一个变量，变量名为HTTP_PROXY，变量值为http://127.0.0.1:1234 方法二： 管理员权限运行CMD： setx HTTP_PROXY http://127.0.0.1:1234

**感谢指点!** @Y0J1G3N

**Comment by zhongfly** [comment]
针对这种需要写了一个脚本，每次mpv启动时，脚本就会获取windows的系统代理,自动给mpv以及ytdl设置代理
https://gist.github.com/zhongfly/2a4e47b844edc22d0cc816f751fe9128

**不需要**设置好ytdl或者mpv代理，脚本都会自动设置。如果发现需要手动设置，可能是其他问题导致
备注：mpv只支持http代理，不支持PAC自动代理脚本！！！
备注2：ytdl还支持socks5代理，但windows系统代理不支持，不在脚本的处理范围之内。

2025.3.1 更新，建议所有使用者都更新！
不再阻塞mpv的启动，之前的版本在我的电脑上会使mpv启动慢5s左右，目前是异步执行，并且优化了性能，只调用一次子进程，优化十分明显。

**Reply by Lonely233233** [reply]
这个脚本应该如何使用

**Reply by zhongfly** [reply]
> 这个脚本应该如何使用

就和其他mpv脚本一样用啊 https://hooke007.github.io/unofficial/mpv_start.html#id14
配置文件夹里scripts/ 自己新建一个名字`xxxxx.lua`(xxxx指脚本名称，用户自定义命名，建议英文)的文件，把源码复制进去保存

__本脚本不保证提供任何售后，不对任何使用后果负责。__

**Reply by zhongfly** [reply]
2025.3.1 更新，建议所有使用本脚本得用户都更新！
不再阻塞mpv的启动，之前的版本在我的电脑上会使mpv启动慢5s左右，目前是异步执行，并且优化了性能，只调用一次子进程，优化十分明显。

终于找到mpv启动慢的原因了

**Reply by hooke007** [reply]
`mp.set_property_native("options/ytdl-raw-options", raw_options)`
这样不是把别的ytdl-raw-options选项给cover掉了吗

**Reply by zhongfly** [reply]
> `mp.set_property_native("options/ytdl-raw-options", raw_options)` 这样不是把别的ytdl-raw-options选项给cover掉了吗

没有吧？
先获取了所有设置
local raw_options = mp.get_property_native("options/ytdl-raw-options")
然后再增加/替换掉了proxy这一项的设置
raw_options["proxy"] = proxy

**Reply by hooke007** [reply]
看了一眼应该不用检查，直接append进去就可以

**Reply by verygoodlee** [reply]
确实用change-list append应该更好
`mp.commandv('no-osd', 'change-list', 'ytdl-raw-options', 'append', 'proxy='..proxy)`

**Comment by eko5624** [comment]
我在mac上用的是ClashX科学上网。也安装了yt-dlp，并加入了环境变量。怎么用mpv来看youtube上的视频。网络代理怎么通过yt-dlp设置呢？

**Reply by czqhurricnae** [reply]
我的情况和你一样，Google 了一圈没有找到方案。

**Reply by Lonely233233** [reply]
只能直接开启tun

**Reply by ghost** [reply]
不仅yt-dlp会自动用系统代理， 但是 ffmpeg 不行，你得手动给 ffmpeg添加代理

**Reply by MiKoto-Railgun** [reply]
我用的这个好像可以看
```
[YouTube_ytdl]
 profile-desc=YouTube-ytdl
 profile-cond=path:match("youtube%.com/")~=nil
 profile-restore=copy
 ytdl-format=bestvideo[height<=?1080][vcodec!=?av01]+bestaudio/best
 stream-lavf-o-append=http_proxy=http://127.0.0.1:7897
 ytdl-raw-options-append=proxy=http://127.0.0.1:7897
```

**Comment by marklove5102** [comment]
感谢提供链接的朋友减少了我搜索 配置的时间.  :)
````txt
mpv.exe --user-agent="Mozilla/5.0 (Linux; Android 6.0; Nexus 5 Build/MRA58N) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/129.0.0.0 Mobile Safari/537.36 Edg/129.0.0.0" --http-header-fields="referer: https://mjv003.com/","origin: https://mjv003.com" --http-proxy=http://192.168.56.103:20174 https://cdnhc.streamfastpro.com/b77765203908516336fcf88e65af4edf/2/2/rDIY76ukDHkj6GDhkgQeXg/rDIY76ukDHkj6GDhkgQeXg.m3u8
````
本机无网  上代理才能上网 播放没问题. 走的是代理.火绒防火墙出现了 mpv.exe 链接代理地址要求过认证弹窗.


![image](https://github.com/user-attachments/assets/8304caba-dcae-4b1f-8476-586707150845)

**Comment by ljxfstorm** [comment]
> --http-proxy=<proxy>
> URL of the HTTP/HTTPS proxy. If this is set, the http_proxy environment is ignored. The no_proxy environment variable is still respected. This option is silently ignored if it does not start with http://. Proxies are not used for https URLs. Setting this option does not try to make the ytdl script use the proxy.

mpv自身提供的http-proxy选项不会应用于HTTPS的URL、也不会传给ytdl，基本上不可用…
对于不想设置环境变量的场景，可以参考如下配置：
```conf
# 传给ytdl的选项，用于基于视频网站URL解析视音频流的源URL，支持SOCKS5协议
ytdl-raw-options-append = proxy=socks5h://127.0.0.1:10808
# 传给ffmpeg的选项，用于从ytdl解析出的源URL加载实际视音频流，只支持HTTP协议
stream-lavf-o-append = http_proxy=http://127.0.0.1:10808
```
